sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -415,31 +414,24 @@ public String getEventHubName() {
             messages.add(message);
         }
 
-        final Context finalSharedContext = sharedContext != null ? sharedContext : Context.NONE;
+        if (isTracingEnabled) {
+            final Context finalSharedContext = sharedContext == null
+                ? Context.NONE
+                : sharedContext.addData(ENTITY_PATH_KEY, eventHubName).addData(HOST_NAME_KEY, fullyQualifiedNamespace);
+            // Start send span and store updated context
+            parentContext.set(tracerProvider.startSpan(finalSharedContext, ProcessKind.SEND));
+        }
 
-        return withRetry(
-            getSendLink(batch.getPartitionId()).flatMap(link -> {
-                if (isTracingEnabled) {
-                    Context entityContext = finalSharedContext.addData(ENTITY_PATH_KEY, link.getEntityPath());
-                    // Start send span and store updated context
-                    parentContext.set(tracerProvider.startSpan(
-                        entityContext.addData(HOST_NAME_KEY, link.getHostname()), ProcessKind.SEND));
-                }
-                return messages.size() == 1
+        return withRetry(getSendLink(batch.getPartitionId())
+            .flatMap(link ->
+                messages.size() == 1
                     ? link.send(messages.get(0))
-                    : link.send(messages);
-
-            })
+                    : link.send(messages)), retryOptions.getTryTimeout(), retryPolicy)
             .doOnEach(signal -> {
```

Comment:
is it called on each try? will it end the span after first retry?

Reply:
No, so with the updated code, it nows does `doOnEach` after the `withRetry` has completed.
So, once withRetry completes either with `error` or `success` , it will send a signal and the `endSpan` uses the signal type to set error or success.